Sparkline memory improvements

matemaciek opened the pull request:
- use cyclic buffer for storing values to reduce fragmentation
- draw as a single `Polygon` instead of multiple `Line`s

In my case current implemntation was running out of memory on ~64 data points, now I can show all 160 data points that I wanted to.

matemaciek committed:
Change sparkline to use cyclic buffer

Since we specify the maximum ammount of data points, we can use cyclic buffer underneath, therefore avoiding memory fragmentation.

This should also help for problems decribed in https://github.com/adafruit/Adafruit_CircuitPython_Display_Shapes/issues/25

matemaciek committed:
Add option to draw polylines (not closed polygons)

matemaciek committed:
Remove myself from authors

Apparently contributors are not listed there

matemaciek committed:
Update polygon.py

Update docs

matemaciek committed:
Change Sparkline to use polygon

Previous implementation used multiple Line elements, causing unnecessary memory overhead

matemaciek commented:
This should help for #25

matemaciek committed:
Add docstrings to new methods

matemaciek committed:
Update sparkline.py

Formatting fixes

matemaciek committed:
Remove unused color from palette

matemaciek committed:
Fix CI errors

Triangle is a subclass of Polygon, so we need to provide support for more than one color in palette

matemaciek committed:
Fix bug spotted in self-review

matemaciek committed:
Use color index constants in triangle

matemaciek committed:
Refactor polygon.py

Decouple polygon drawing from creating bitmap, so the code can be reused for drawing on external bitmaps

matemaciek committed:
Pylint fixes

Fix some of the old warnings, reenable some of overrides

matemaciek committed:
Reuse the same bitmap in sparkline

matemaciek committed:
Add MultiSparkline

And rewrite Sparkline to be a subclass of MultiSparkline

matemaciek committed:
Fix/override pylint remarks

matemaciek committed:
One more pylint remark

FoamyGuy commented on adafruit_display_shapes/multisparkline.py:
```diff
@@ -0,0 +1,337 @@
+                self.y_bottoms[i] = bottom
+
+                if update:
+                    self.update(i)
```
Is this supposed to be `self.update_line(i)`? 

I attempted to run the simpletest from the examples folder with the modified version of the library from this PR and end up with an error leading to this line:
```
code.py output:
Traceback (most recent call last):
  File "code.py", line 131, in <module>
  File "/lib/adafruit_display_shapes/sparkline.py", line 96, in add_value
  File "/lib/adafruit_display_shapes/multisparkline.py", line 197, in add_values
TypeError: function takes 1 positional arguments but 2 were given
```

As far as I can tell MultiSparkline does not "normally" have an `update()` function. But SparkLine does extend MultiSparkline so if the object is a SparkLine then it will have an `update()` function, but not otherwise. And in the case when it does have the `update()` function it has no arguments other than self, so `i` should not get passed in to that one.

matemaciek replied:
Yes, it was previously overloaded method `update`, but after pylint complained I chenged it to `update_line` and forgot about this place. I'll fix it (-:

FoamyGuy commented on adafruit_display_shapes/sparkline.py:
```diff
@@ -79,36 +76,11 @@ def __init__(
-        self.y_bottom = y_min
-        # y_bottom: The actual minimum value of the vertical scale, will be
-        # updated if autorange
-        self.y_top = y_max
```
`y_top` and `y_bottom` are used by the `display_shapes_sparkline_ticks.py` example that is in this repo. That example is unable to run with the version of code from this PR since those have been removed. 

I think we can add them back by making properties inside of sparkline.py like this:
```
    @property
    def y_top(self):
        return self.y_tops[0]

    @property
    def y_bottom(self):
        return self.y_bottoms[0]
```

I tested this out and it does seem to allow that example to run successfully.

matemaciek replied:
Agreed, I'll add them. I'll also add it for `y_min` and `y_max`, as they might be considered public.

matemaciek replied:
...or maybe let's not expose `y_min` and `y_max`? They have no advantage over `y_bottom` and `y_top` and might be misleading.

FoamyGuy reviewed (changes requested):
@matemaciek Thanks for working on improving the efficiency of this!

I've spotted a few issues attempting to run the existing examples with the new version of the repo. We'll definitely want to get those straightened out either by modifying the classes as needed, or modifying the examples if an API has changed or something causing a need for that. 

Some are noted at relevant sections of the code. 

With the "triple" example, specifically sparkline2 within it, I've also found that the new version of the library raises a pixel out of bounds error:
```
Traceback (most recent call last):
  File "code.py", line 245, in <module>
  File "/lib/adafruit_display_shapes/sparkline.py", line 96, in add_value
  File "/lib/adafruit_display_shapes/multisparkline.py", line 197, in add_values
  File "/lib/adafruit_display_shapes/multisparkline.py", line 316, in update_line
  File "/lib/adafruit_display_shapes/multisparkline.py", line 237, in _draw
  File "/lib/adafruit_display_shapes/polygon.py", line 101, in draw
  File "/lib/adafruit_display_shapes/polygon.py", line 161, in _line_on
IndexError: pixel coordinates out of bounds
```

It seems to only be a problem with sparkline2, but I'm not certain exactly why that one is the problematic one. Likely something about the behavior around the range, or size and shape has changed a little bit causing it.

I've started debugging a bit, but don't have a conclusive answer yet. So far it's looking like it tries to put pixels either at negative y values, or at y value 100 which is 1 pixel too high, 99 is the largest available index in the bitmap in question. 

From that it seems to point toward there being some logic issue in the conversion between points in the range of the sparkline graph to points in the 2D plane of the pixels in the bitmap. Or maybe a bug in the example that was previously handled more gracefully despite being a bug? 

We'll need to get it figured out and fixed as well though. We want to make sure all examples in the repo. are able to run successfully.

matemaciek commented:
Thanks for the review! I've addressed inline comments directly. As for of-by-one errors: The existing version (in original repo) seem to have a semi-bug where in some scenarios it draws on larger bitmap than requested - but since bitmap size is auto-set basing on list of points provided, they always fit. I'll debug the existing example that is failing, as I tought I fixed that bug (-:

matemaciek commented:
While I'm working on fixing issues above, a screenshot of how it looks in action:
![20230107_201345](https://user-images.githubusercontent.com/423299/211166918-5495dcfa-2594-438d-8894-9144c407cd6a.JPG)

matemaciek committed:
Address review remarks

matemaciek commented:
OK, I know where the bug is (-: When reorganising the code I broke an edge case when first point of sparkline would be out of screen. Working on a fix.

matemaciek committed:
Move checking if line fits on bitmap to polygon class

matemaciek committed:
Pylint compliance

FoamyGuy commented on adafruit_display_shapes/polygon.py:
```diff
@@ -33,121 +33,162 @@
 
-    # pylint: disable=invalid-name, too-many-locals, too-many-branches
+    @staticmethod
+    def draw(
```
I'm curious about a few of these methods being static. `draw()`, `_line_on()` and `_safe_draw()`  Did this change to static factor in to memory efficiency somehow?

Is there a situation in which they are intended for use outside of in an instance of this class ultimately inside of a sparkline? 

My first inclination was that they could be regular functions and could access bitmap from self rather than having it passed in separately. That way they are always acting upon only the bitmap that is part of this instance.

matemaciek replied:
I see resposibilities as follows:
- `Polygon` class: class containing logic of drawing polygons
- `polygon` object: bitmap containing single drawn polygon

Having `draw()` method as static alows us to reuse drawing logic to draw multiple times on the same bitmap - which allows for two things previously not possible:
- having multiple polygons on single bitmap (lower memory usage)
- reusing same bitmap for sequential redrawing of different polygons on the same bitmap (common case for sparklines, and not recreating bitmap over and over potentially means less fragmentation).

FoamyGuy commented on adafruit_display_shapes/polygon.py:
```diff
@@ -33,121 +33,162 @@
+
+    # pylint: disable=too-many-branches, too-many-locals
+    @staticmethod
+    def _line_on(
```
I think it may be possible to use this built-in bitmaptools function from the core rather than having the implementation in python here. https://docs.circuitpython.org/en/latest/shared-bindings/bitmaptools/index.html#bitmaptools.draw_line

Perhaps ultimately we could make a `draw_polygon()` function inside bitmaptools as well which could then be used generally from anywhere, including this library.

matemaciek replied:
It's great to see that built-in function, I didn't knew it exists! Feel free to approve/merge this MR already, I'll work on using built-in function either way - one doubt I have is handling out-of-bonds coordinates by it, I'll need to check how it behaves.

As for adding `draw_polygon()` to bitmaptools I'm totally up for it, sounds like a nice starting point for contribting to circuitpython core (-:

matemaciek replied:
I just checked, `bitmaptools.draw_line()` does not like lines going out of bitmap - I think that it could be a valid scenario, so I may end up trying to change that behaviour in `bitmaptools` before using it here.

FoamyGuy reviewed (commented):
Thanks for digging in and figuring out some fixes!

I think you also fixed a another issue I noticed yesterday (with the released version ) where the dynxwidth config sometimes causes the line to be drawn below the chart. The latest version from this PR does not exhibit that issue any more. 

The functionality on this is looking good to me now. I tested the 3 existing examples and all are looking good with this version. 

I have one question about the static functions, and one question whether we can use a builtin draw_line instead of needing an implementation here.

FoamyGuy commented:
> While I'm working on fixing issues above, a screenshot of how it looks in action: ![20230107_201345](https://user-images.githubusercontent.com/423299/211166918-5495dcfa-2594-438d-8894-9144c407cd6a.JPG)

That looks awesome! Thank you for sharing! Do post anywhere online about your project? If you interested it could totally be submitted for inclusion in the weekly python on hardware newsletter.

matemaciek commented:
> That looks awesome! Thank you for sharing! Do post anywhere online about your project? If you interested it could totally be submitted for inclusion in the weekly python on hardware newsletter.

Thanks! I'm definitely up for including my project in the newsletter (-: I think it would be best to do it after I publish code of it, so If someone is interested, they can reuse it. My goal was to build stand-alone CO₂ sensor, so one can know when to open the window and let more air in (just by color-coding values). I ended up adding screen, then some in-memory logging, then simple webserver for better viewing of charts, all living on Pico W (I try to be cost-aware in this project). Without improvements it was running out of memory after ~30 data points, now it seem to be safely running on 160 data points while serving some data online, but I want to fit more (-:

FoamyGuy reviewed (approved):
Looks good to me. Thank you!
